Update `dense_vector` docs with kNN indexing options (#80306)
elastic/elasticsearch: docs/reference/mapping/types/dense-vector.asciidoc

<titleabbrev>Dense vector</titleabbrev>
++++

The `dense_vector` field type stores dense vectors of float values.

You can use `dense_vector` fields in
<<query-dsl-script-score-query,`script_score`>> queries to score documents.

> They can also be indexed to support efficient k-nearest neighbor search. Dense

Author: I'll add a link to the `_knn_search` endpoint here once I write the docs for that (in a follow-up PR).

> They can also be indexed to support efficient k-nearest neighbor search. Dense

Reviewer: > k-nearest neighbor search
Note from @jtibshirani: Link this to the knn search API docs when they're up (in a follow-up PR).

Sorry for accidentally deleting this. I was deleting some of my own comments from a pending review and accidentally clicked the wrong one. 🤦

perform a brute-force kNN search by scanning all documents and ranking them by
similarity.

In many cases, a brute-force kNN search is not efficient enough. For this
reason, the `dense_vector` type supports indexing vectors into a specialized
data structure to support fast kNN search. You can enable indexing by setting
through the `index` parameter:

[source,console]
--------------------------------------------------
PUT my-index-000002
{
  "mappings": {
    "properties": {
      "my_vector": {
        "type": "dense_vector",
        "dims": 3,
        "index": true,
        "similarity": "dot_product" <1>
      }
    }
  }
}
--------------------------------------------------
<1> When `index` is enabled, you must define the vector similarity to use in kNN search

Elasticsearch uses the https://arxiv.org/abs/1603.09320[HNSW algorithm] to
support efficient kNN search. Like most kNN algorithms, HNSW is an approximate
method that sacrifices result accuracy for improved speed.

NOTE: indexing vectors for kNN search is an expensive process. It can take
substantial time to ingest documents that contain vector fields with `index`
enabled.

[discrete]
[[dense-vector-params]]

> ==== Parameters for dense vector fields

Author: I struggled with the formatting a bit here -- any corrections/ suggestions are appreciated. The other field types use the `[horizontal]` styling, but I couldn't get this to work with the sublists and code snippets.

Author: Thanks! I went through and accepted all of these (batched into a new commit).

}

--------------------------------------------------
<1> `dims` is a required parameter that defines the number of vector dimensions

> NOTE:  Unlike most other data types, dense vectors are always single-valued.

Reviewer: Nit: whitespace
```suggestion
NOTE: Unlike most other data types, dense vectors are always single-valued.
```

It is not possible to store multiple values in one `dense_vector` field.

> ==== Indexing vectors for kNN search

Reviewer: I'd add an anchor here.
```suggestion
[[index-vectors-knn-search]]
==== Index vectors for kNN search
```

experimental[]

_k-nearest neighbor search_ or _kNN_ is the task of finding the k nearest
vectors to a query vector, as measured by a similarity metric.

As mentioned above, `dense_vector` fields can be used to rank documents
through <<query-dsl-script-score-query,`script_score` queries>>. This lets you
perform a brute-force kNN search by scanning all documents and ranking them by
similarity.

In many cases, a brute-force kNN search is not efficient enough. For this
reason, the `dense_vector` type supports indexing vectors into a specialized
data structure to support fast kNN search. You can enable indexing by setting

> through the `index` parameter:

Author: Thanks, this is great content! I was purposefully quite concise here, because I didn't think the field type reference was the right place to explain this. What would you think about saving this content for the high-level guide, and making sure to link to it here? I am not worried that we'll forget to add the content, we have plenty of time before 8.0 and we're tracking it as a requirement.

> through the `index` parameter:

Reviewer: That makes sense to me. I'll go ahead and mark this as resolved.

[source,console]
--------------------------------------------------
PUT my-index-000002
{
  "mappings": {
    "properties": {
      "my_vector": {
        "type": "dense_vector",
        "dims": 3,
        "index": true,
        "similarity": "dot_product" <1>
      }
    }
  }
}
--------------------------------------------------
<1> When `index` is enabled, you must define the vector similarity to use in kNN search

Elasticsearch uses the https://arxiv.org/abs/1603.09320[HNSW algorithm] to
support efficient kNN search. Like most kNN algorithms, HNSW is an approximate

> method that sacrifices result accuracy for improved speed.

Author: I'm curious if there's a guideline/ philosophy underlying this -- do people tend not to read past snippets and miss the information? To me it feels natural to say "You can enable indexing through the
`index` parameter:" then immediately show the snippet.

> method that sacrifices result accuracy for improved speed.

Reviewer: I didn't have a particular guideline or philosophy in mind. My thinking was that HNSW info is only applicable if `index` is `true`, and that might be good to know before I create a mapping.

However, I don't feel strongly about it. I'm good with leaving it here, particularly since we want to keep the intro to this section concise.

NOTE: indexing vectors for kNN search is an expensive process. It can take
substantial time to ingest documents that contain vector fields with `index`
enabled.

> [discrete]

Author: For my knowledge, what does `[role="child_attributes"]` do?

> [discrete]

Reviewer: It adds a border and other styling for collapsible sections that contain a nested definition list. Here's an example from the search API docs:

<img width="750" alt="Screen Shot 2021-11-04 at 1 30 20 PM" src="https://user-images.githubusercontent.com/40268737/140389771-a03e211d-9e47-4eff-941a-25a27fccc988.PNG">

[[dense-vector-params]]
==== Parameters for dense vector fields

The following mapping parameters are accepted:

> `dims`::
> 
> The number of vector dimensions. This value is required and cannot exceed 2048.

Reviewer: ```suggestion
`dims`::
(Required, integer)
Number of vector dimensions. Can't exceed `2048`.
```

> `index`::
> 
> Whether the vectors should be indexed to support efficient k-nearest neighbor search. Defaults to `false`.

Reviewer: ```suggestion
`index`::
(Optional, Boolean)
If `true`, you can search this field using the kNN search API. Defaults to
`false`.
```

The vector similarity metric to use in kNN search. Documents are ranked by
their vector field's similarity to the query vector. The `_score` of each
document will be derived from the similarity, in a way that ensures scores are
positive and that a larger score corresponds to a higher ranking.
+
When `index` is enabled, this parameter is required. The following options are supported:

`l2_norm`:::
Computes similarity based on the L^2^ distance (also known as Euclidean
distance) between the vectors. The document `_score` is computed as
`1 / (1 + l2_norm(query, vector)^2)`.`

`dot_product`:::
Computes the dot product of two vectors. This option provides an optimized way
to perform cosine similarity. In order to use it, all vectors must be of unit
length, including both document and query vectors. The document `_score` is
computed as `(1 + dot_product(query, vector)) / 2`.

`cosine`:::
Computes the cosine similarity. Note that the most efficient way to perform
cosine similarity is to normalize all vectors to unit length, and instead use
`dot_product`. You should only use `cosine` if you need to preserve the
original vectors and cannot normalize them in advance. The document `_score`

> is computed as `(1 + dot_product(query, vector)) / 2`.

Author: 👍

NOTE: Although they are conceptually related, the `similarity` parameter is

> different from <<text,`text`>> field <<similarity,`similarity`>> and accepts

Reviewer: Should we mention keyword fields too?

Author: It's actually pretty strange (and I assume rare) to set `similarity` on a `keyword` field. I think this would be more distracting than helpful. We've actually considered removing support for `similarity` on `keyword`.

a distinct set of options.

`index_options`::

An optional section that configures the kNN indexing algorithm. The HNSW
algorithm has two internal parameters that influence how the data structure is
built. These can be adjusted to improve the accuracy of results, at the

> expense of slower indexing speed. The `index_options` object contains the
> following properties:
> 
> `type`:::

Reviewer: Adds a collapsible section + required and data type.
```suggestion
expense of slower indexing speed. 
+
.Properties of `index_options`
[%collapsible%open]
====
`type`:::
(Required, string)
```

The type of kNN algorithm to use. Currently only `hnsw` is supported.

> `m`:::

Reviewer: Adds required and data type.
```suggestion
`m`:::
(Required, integer)
```

> Controls how many of the nearest neighbor candidates are connected to each new

Reviewer 2: may be better say, the maximum number of neighbour connections each node can have in the HNSW graph?

Author: 👍  I reworded this to make it clearer. But I didn't say "maximum" because this is the actual number of connections that the algorithm tries to create (not an upper bound).

Reviewer 2: My interpretation was that it is an upper bound, in the sense that we don't allow to have more neighbours than `m`, and will prune the worst neighbours to have the number of neighbours always under `m`.

Author: Oh I see, I interpreted "maximum" to mean that it's possible many nodes have fewer than `m` neighbors. But the algorithm attempts to have each node have exactly `m` neighbors.

In any case, hopefully the current wording is clear :) It matches the [description from hnswlib](https://github.com/nmslib/hnswlib/blob/master/ALGO_PARAMS.md): "M - the number of bi-directional links created for every new element during construction".

their vector field's similarity to the query vector. The `_score` of each
document will be derived from the similarity, in a way that ensures scores are
positive and that a larger score corresponds to a higher ranking.
+
^*^ If `index` is `true`, this parameter is required.
+
.Valid values for `similarity`
[%collapsible%open]
====
`l2_norm`:::
Computes similarity based on the L^2^ distance (also known as Euclidean
distance) between the vectors. The document `_score` is computed as
`1 / (1 + l2_norm(query, vector)^2)`.`

`dot_product`:::
Computes the dot product of two vectors. This option provides an optimized way
to perform cosine similarity. In order to use it, all vectors must be of unit
length, including both document and query vectors. The document `_score` is
computed as `(1 + dot_product(query, vector)) / 2`.

`cosine`:::
Computes the cosine similarity. Note that the most efficient way to perform
cosine similarity is to normalize all vectors to unit length, and instead use
`dot_product`. You should only use `cosine` if you need to preserve the
original vectors and cannot normalize them in advance. The document `_score`
is computed as `(1 + cosine(query, vector)) / 2`.

> 

Reviewer: I think this got lost (or I forgot to suggest it), but this closes the collapsible section.
```suggestion
====
+
```
